runtime/src/utils/buffer/append.rs

```diff
@@ -0,0 +1,395 @@
+        })
+    }
+
+    /// Change the capacity of the write buffer.
+    ///
+    /// The buffer will be flushed, leaving an empty buffer upon return.
+    pub async fn reset_buffer(&mut self, buffer_size: usize) -> Result<(), Error> {
+        let mut buffer = self.buffer.write().await;
+        self.flush(&mut buffer).await?;
+        buffer.capacity = buffer_size.max(self.pool_ref.page_size * 2);
+
+        Ok(())
+    }
+
+    /// Append all bytes in `buf` to the tip of the blob.
+    pub async fn append(&self, buf: impl Into<StableBuf> + Send) -> Result<(), Error> {
+        // Prepare `buf` to be written.
+        let buf = buf.into();
+
+        // Acquire a write lock on the buffer.
+        let mut buffer = self.buffer.write().await;
+
+        // Ensure the write doesn't overflow.
+        buffer
+            .size()
+            .checked_add(buf.len() as u64)
+            .ok_or(Error::OffsetOverflow)?;
+
+        if buffer.append(buf.as_ref()) {
+            // Buffer is over capacity, flush it to the underlying blob.
+            return self.flush(&mut buffer).await;
+        }
+
+        Ok(())
+    }
+
+    /// Returns the current logical size of the blob including any buffered data.
+    ///
+    /// This represents the total size of data that would be present after flushing.
+    #[allow(clippy::len_without_is_empty)]
+    pub async fn size(&self) -> u64 {
+        let buffer = self.buffer.read().await;
+        buffer.size()
+    }
+
+    /// Flush the append buffer to the underlying blob, caching each page worth of written data in
+    /// the buffer pool.
+    async fn flush(&self, buffer: &mut Buffer) -> Result<(), Error> {
+        // Take the buffered data, if any.
+        let Some((buf, offset)) = buffer.take() else {
+            return Ok(());
+        };
+
+        // Insert the flushed data into the buffer pool. This step isn't just to ensure recently
+        // written data remains cached for future reads, but is in fact required to purge
+        // potentially stale cache data which might result from the edge the case of rewinding a
+        // blob across a page boundary.
+        let remaining = self.pool_ref.cache(self.id, &buf, offset).await;
+
+        // If there's any data left over that doesn't constitute an entire page, re-buffer it into
+        // the append buffer to maintain its page-boundary alignment.
+        if remaining != 0 {
+            buffer.offset -= remaining as u64;
+            buffer.data.extend_from_slice(&buf[buf.len() - remaining..])
+        }
+
+        // Write the data buffer to the underlying blob. Note that the implementation will rewrite
+        // the last (blob_size % page_size) "trailing bytes" of the underlying blob since the
+        // write's starting offset is always page aligned. We don't expect this inefficiency to be a
+        // significant performance concern, but would be easy enough to avoid by maintaining the
+        // underlying blob's size.
```

Comment:
I consider this "future work" enough that we should link an issue here instead of leaving a dangling suggestion.

Reply:
added TODO w/ pointer to an issue.